Base work on IDDFS

felix91gr opened the pull request:
Done:
- Added `iddfs`
- Added docs based on `dfs` docs
- Adapted `idastar` doctests
- Added it to the `directed` module and its API to the library `prelude`.

Close #132 #134.

felix91gr commented:
bors r+

felix91gr commented:
(I wonder if I can ask bors to do a test, it'll probably not hear my command now that I think of it :sweat_smile:)

samueltardieu commented on src/directed/iddfs.rs:
```diff
@@ -0,0 +1,125 @@
+        let mut best_result = Path::Impossible;
+
+        for n in neighbours_it {
+            let mut still_possible_from_this_node = !path.contains(&n);
```
I think this mutable variable is not needed see the next comments.

felix91gr replied:
I think I agree! Thank you, I'll try it tomorrow when I'm at the computer again 😊

samueltardieu commented on src/directed/iddfs.rs:
```diff
@@ -0,0 +1,125 @@
+        for n in neighbours_it {
+            let mut still_possible_from_this_node = !path.contains(&n);
+
+            if still_possible_from_this_node {
```
You can use `if !path.contains(&n)` instead.

samueltardieu commented on src/directed/iddfs.rs:
```diff
@@ -0,0 +1,125 @@
+                path.push(n);
+                match step(path, neighbours, success, depth - 1) {
+                    Path::FoundOne => return Path::FoundOne,
+                    Path::NoneAtThisDepth => (),
```
You can use `Path::NoneAtThisDepth => best_result = Path::NoneAtThisDepth`.

samueltardieu commented on src/directed/iddfs.rs:
```diff
@@ -0,0 +1,125 @@
+                match step(path, neighbours, success, depth - 1) {
+                    Path::FoundOne => return Path::FoundOne,
+                    Path::NoneAtThisDepth => (),
+                    Path::Impossible => still_possible_from_this_node = false,
```
And `Path::Impossible => ()`

samueltardieu commented on src/directed/iddfs.rs:
```diff
@@ -0,0 +1,125 @@
+
+            if still_possible_from_this_node {
+                best_result = Path::NoneAtThisDepth;
+            }
```
And then this block can be removed.

samueltardieu commented:
The tests and benchmark are located in the `tests` and `benches` directories. You should add a test every time there is a `bfs` test. Also, `README.md` should be updated.

samueltardieu commented:
bors try

samueltardieu commented on src/directed/iddfs.rs:
```diff
@@ -0,0 +1,125 @@
+{
+    if depth == 0 {
+        Path::NoneAtThisDepth
+    }
```
This block should be removed as this can never happen: `step` is an internal function and the smallest value we pass to `depth` is 1.

felix91gr replied:
No, it always reaches 0. 0 is the base case of the recursion, and every call reduces `depth` by one.

I should be naming it "remaining_depth" or something like that, that would be way more descriptive I think

samueltardieu replied:
Indeed you're right, I forgot about the `- 1` when recursing of course.

felix91gr commented:
> The tests and benchmark are located in the tests and benches directories. You should add a test every time there is a bfs test. Also, README.md should be updated.

I see (and indeed, I totally forgot about the README 😅)

felix91gr commented:
I'm having really bad results with the benchmarks. Maybe it's because I'm using heap allocations somewhere.

:thinking: 

`path.push()` and `path.pop` don't count, or shouldn't count. They are `O(1)` most of the time, except when the vector duplicates itself. Since I'm using always the same vector, the number of duplications is log(N), where N is the size of the shortest path. Hmm.

The problem might be that `neighbors` allocates a new `Vec` every time it's called. Is there a way to define it lazily, such that I can use the stack instead? I'm going to investigate further.

felix91gr commented:
If it's way slower than BFS, that is because for problems where you can reopen many already-seen states, IDDFS makes many redundant openings of states. One way to solve this is to use a hash table where you'd store states that you've already seen. But that takes away the linear space guarantee of IDDFS.

I guess it's good to have, but maybe only when you just can't afford the memory usage of BFS and don't have a heuristic to support IDA*.

felix91gr commented:
Do you know how I can implement this?

> I'm going to write a benchmark to see how many nodes IDDFS is opening on every depth level v/s how many BFS is opening there.

If this was idk, C or Python I'd be using a global variable. But I don't think I can do that in Rust (which I think is a good thing).

felix91gr commented:
Okay I made it work. These are the results:

```console
running 2 tests
BFS opened 12 nodes in a 2 x 2 board
BFS opened 12 nodes in a 3 x 3 board
BFS opened 16 nodes in a 4 x 4 board
BFS opened 25 nodes in a 5 x 5 board
test bfs_opened_bench          ... bench:       6,156 ns/iter (+/- 737)
IDDFS opened 280 nodes in a 2 x 2 board
IDDFS opened 1174 nodes in a 3 x 3 board
IDDFS opened 10765 nodes in a 4 x 4 board
IDDFS opened 791015 nodes in a 5 x 5 board
test iddfs_opened_bench        ... bench: 122,228,212 ns/iter (+/- 6,782,084)
```

felix91gr commented:
Also, a friend made me remember that iddfs with a hashtable is usually faster than bfs when you have a very-connected state space. 

Regarding space, the hashtable's size is bounded by the space that bfs would use. But you only really reach that size when you're exploring at the optimum's depth, and since iddfs goes depth-first, the chance of finding the optimum before having the hashtable cover the entire depth-bounded space is very high.

TL;DR: I think the performance issue is just a property of iddfs. I forgot how much speed you sacrifice without the hashtable. Should I give it a hashtable, or should I leave it like it is atm?

felix91gr commented:
@samueltardieu Should I finish this PR?

In performance:

1. It guarantees linear memory usage
2, It can be very slow depending on the layout of the state space (although it's still exponential in the worst case like BFS)

samueltardieu commented on benches/algos.rs:
```diff
@@ -3,7 +3,7 @@
 extern crate test;
 
-use pathfinding::prelude::{astar, bfs, dfs, dijkstra, fringe, idastar};
+use pathfinding::prelude::{astar, bfs, dfs, iddfs, dijkstra, fringe, idastar};
```
That's not in the right order.

felix91gr replied:
Oh! I'm using that ordering since it made sense from a conceptual standpoint. But I see now that you're using an alphabetical ordering? I'll follow suit then :)

samueltardieu commented on src/directed/iddfs.rs:
```diff
@@ -0,0 +1,119 @@
+    FoundOne,
+    NoneAtThisDepth,
+    Impossible,
+}
```
Any reason not to put those in analphabetical order?

felix91gr replied:
I think that ordering was conceptual:
* Either you've finished searching,
* You're still searching,
* Or you've concluded there's no solution

felix91gr replied:
I can totally make it alphabetical though :)

samueltardieu commented on src/directed/mod.rs:
```diff
@@ -3,6 +3,7 @@
 pub mod astar;
 pub mod bfs;
 pub mod dfs;
+pub mod iddfs;
```
Not ordered properly.

samueltardieu commented on src/lib.rs:
```diff
@@ -19,6 +19,7 @@ pub mod prelude {
     pub use directed::astar::*;
     pub use directed::bfs::*;
     pub use directed::dfs::*;
+    pub use directed::iddfs::*;
```
Not ordered properly.

samueltardieu commented:
Sure! I've added some stylistic comments. Also, I think the git commit messages should be reworked to be more "canonical" (no smileys, no long first line, no ":" on the first line which should be self-sufficient). I'll give bors a try but I think this will block because of style considerations at this stage.

bors try

felix91gr commented:
Okay, I've completed (I think) the PR. I have to squash the commits now.

Take a look at the README, I have two different wordings for IDDFS:

- https://github.com/felix91gr/pathfinding/blob/c67a2533b62afd2fe28ea910ea8992339a06873b/README.md
- https://github.com/felix91gr/pathfinding/blob/4c4f97fdd464ed4a186e22b31c8de845693daf35/README.md

Which one do you think is best? I tried making it shorter, but it still uses more than one line in my browser. So idk really

felix91gr commented:
> but please just use "IDDFS" as the entry name in the list, not the expanded name. Just like this is done for "IDA*" for example.

Done! I saw how you put DFS there and thought I should do the same :)

I'll squash now.

samueltardieu commented on src/directed/iddfs.rs:
```diff
@@ -0,0 +1,119 @@
+    }
+    else if success(path.last().unwrap()) {
+        Path::FoundOptimum
+    }
```
Note that you should use `rustfmt` or the bot will still refuse the patch. For example, `else` must be on the same line as `{` as per Rust default style.

felix91gr replied:
Oh, good to know :D

felix91gr commented:
> Did you get cargo fmt from rustup with component rustfmt-preview? This is the latest rustfmt version.

Yes! This is what I've got:

```console
felix@felix-X550LD ~/D/P/R/pathfinding> rustup component add rustfmt-preview
info: downloading component 'rustfmt-preview'
  3.6 MiB /   3.6 MiB (100 %) 529.4 KiB/s ETA:   0 s                
info: installing component 'rustfmt-preview'
felix@felix-X550LD ~/D/P/R/pathfinding> cargo fmt
felix@felix-X550LD ~/D/P/R/pathfinding> git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   benches/algos-fill.rs
	modified:   benches/algos.rs
	modified:   examples/sliding-puzzle.rs
	modified:   src/directed/astar.rs
	modified:   src/directed/edmonds_karp.rs
	modified:   src/directed/iddfs.rs
	modified:   src/undirected/connected_components.rs
	modified:   tests/edmondskarp.rs
	modified:   tests/gps.rs
	modified:   tests/pathfinding.rs

no changes added to commit (use "git add" and/or "git commit -a")

```

> (I'll add this in README.md btw)

The formatting? That would be nice :)
Also: maybe the bot can comment or something saying "You should run `cargo fmt`, because I follow that style of Rust" (not the exact wording but you get what I mean).

samueltardieu commented:
Are you sure it finds the right `cargo-fmt`? Are you using the `nightly` toolchain by default? What does `cargo +nightly fmt` do?

felix91gr commented:
Ahaa, therein lied the problem. This is my output now:

```console
felix@felix-X550LD ~/D/P/R/pathfinding> rustup component add rustfmt-preview --toolchain nightly
info: downloading component 'rustfmt-preview'
  3.9 MiB /   3.9 MiB (100 %) 769.8 KiB/s ETA:   0 s                
info: installing component 'rustfmt-preview'
felix@felix-X550LD ~/D/P/R/pathfinding> cargo +nightly fmt
felix@felix-X550LD ~/D/P/R/pathfinding> git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   benches/algos-fill.rs
	modified:   benches/algos.rs
	modified:   src/directed/iddfs.rs

no changes added to commit (use "git add" and/or "git commit -a")
felix@felix-X550LD ~/D/P/R/pathfinding> 
```

It seems we're in the clear now :D

felix91gr commented:
Okay, `rustfmt` isn't changing anything anymore (I'm using the nightly one now). I think the PR is ready now for bors :)

samueltardieu commented:
Hopefully a last change request: could you please change "Added IDDFS" with "Add IDDFS" in the git commit message? The style used in this repository is "imperative commit messages", where you order the commit to do something ("Merge …", "Add …", "Fix …"), so it requires an imperative verb. I'll add to the README when the other changes are merged too.

Thanks for your patience.

felix91gr commented:
Haha, sure :laughing: 

> Thanks for your patience.

It's okay! Coming from style-free workflows, I feel lost sometimes when I have to adhere to styles because I haven't yet learned the kicks of a style-bound workflow. Thankfully you have had patience with me too, and we have `rustfmt` to help with the code side ^^

samueltardieu commented:
bors r+

samueltardieu commented:
bors r+

samueltardieu commented:
Mmm, the questions from the initial merge requests got pulled in the merge message for Bors. I'll clean up the initial post of this thread and reapprove bors.

bors r-

samueltardieu commented:
bors r+

samueltardieu commented:
bors r+

samueltardieu commented:
Added issues # in description.

bors r-
bors r+
